woocommerce/woocommerce
# Add separator to "Place order" button and use the block style variations API

### Changes proposed in this Pull Request:

Firstly this cherry-picks some changes that @opr implemented in this PR: https://github.com/woocommerce/woocommerce/pull/49252 

Furthermore it closes #48433 by using the block style variations API to allow customization of the separator. 

I'm interpreting from context of the discussions had about solving this problem, basically there was some concern that a purely CSS based solution made it difficult to customize the separator before. While the style variations API is still CSS based it allows a simpler approach for customization of the separator. Implementers can register additional style variations for the actions block and then one line of CSS targeting the new variation class name will allow them to style their new variation. 

It's worth mentioning that ootb the style variation API does not support something like choosing a separator from a list or enetering your own character for the separator. This is possible as a more custom solution using the available APIs, but would be a more complex implementation. So this doesn't give full front-end customization to merchants, but does future proof the customization of the separator for theme authors and 3PDs.

Example:


```php
register_block_style(
  'woocommerce/checkout-actions-block',
    array(
	'name'       => 'triple-dot-separator',
	'label'      => __( '"Triple-Dot" Separator', 'woocommerce' ),
	'is_default' => true,
    )
);
```

```scss
.wp-block-woocommerce-checkout-actions-block.is-style-triple-dot-separator {
	.wc-block-components-checkout-place-order-button__separator {
		&::after {
			content: "···";
		}
	}
}
```

As part of this I added the default separator and a `-` separator. But we could omit the dash separator for now if we wish to, it just serves as an example.

One thing I found a bit odd about the variations API is that the default variation doesn't seem to apply a class name. It sort of makes sense for the standard use-case, but for clarity I think the default could have a class applied. It could be for performance reasons.

I've also added an example to the block settings to allow the style variation previews to work.



### How to test the changes in this Pull Request:

(Copy pasted from #49252)

1. Add an item to your cart and go to the Checkout page.
2. Look at the place order button. It should not have a price or separator visible.
3. Press the button to place your order. Ensure the spinner shows, and then disappears when the checkout is done processing.
4. Go to the Checkout page in the post editor.
5. Select the Checkout Actions Block (The place order button).
6. In the sidebar, toggle the Show price in the button option on.
7. In the editor, look at the place order button. Ensure the price is visible with a separator between the text and the price.
8. Ensure you can edit the text (click into it and type, it's not super obvious it's editable).
9. Try with different lengths of text and ensure the button behaves reasonably, i.e. doesn't grow/shrink too much, content doesn't overflow, the price and text have enough room etc.
10. Save your changes to the text and visit the front end. Repeat steps 1, 2, and 3 ensuring your new text is visible, and that the place order button is still working normally.

Additionally you can test that the block previews work for the provided `default` and `dash` separator:

1. In admin, go to edit the Checkout page
2. Select the "Place order" button near the bottom of the checkout order form
3. Under the "Styles" panel in the inspector, hover over the "Dot" Separator and "Dash" Separator buttons to see respective visual previews of each.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
Add a separator to the Checkout "Place order" button, customizable via block style variations.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
